### ISSUE TITLE
Average IoU suddenly drops to 0 during training !

### ISSUE THREAD

**Issue by nizarmasmoudi** [AlexeyAB/darknet#7281]
I am training a **YOLOv4** algorithm to detect people (1 class) using **VisDrone** Dataset. The training starts normally:
```
 (next mAP calculation at 5903 iterations) 
 3: 5562.074707, 5559.250488 avg loss, 0.000000 rate, 1.716530 seconds, 24 images, 1.030408 hours left
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 139 Avg (IOU: 0.255043), count: 11, class_loss = 12043.081055, iou_loss = 138.799805, total_loss = 12181.880859 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 150 Avg (IOU: 0.396577), count: 23, class_loss = 3606.523682, iou_loss = 225.732666, total_loss = 3832.256348 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 161 Avg (IOU: 0.401758), count: 12, class_loss = 1082.762085, iou_loss = 49.273315, total_loss = 1132.035400 
 total_bbox = 667, rewritten_bbox = 1.349325 % 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 139 Avg (IOU: 0.273875), count: 34, class_loss = 12006.184570, iou_loss = 602.565430, total_loss = 12608.750000 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 150 Avg (IOU: 0.281148), count: 44, class_loss = 3579.320068, iou_loss = 380.955078, total_loss = 3960.275146 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 161 Avg (IOU: 0.246588), count: 10, class_loss = 1087.403564, iou_loss = 27.188965, total_loss = 1114.592529 
 total_bbox = 755, rewritten_bbox = 1.986755 %
...
```
After about 50 minutes in, `Avg IOU` starts "vanishing" and gets stuck at `0`:
```
 (next mAP calculation at 5903 iterations) 
 1268: 18.638748, 64.584755 avg loss, 0.001000 rate, 1.572169 seconds, 10144 images, 0.381671 hours left
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 139 Avg (IOU: 0.000000), count: 1, class_loss = 0.000000, iou_loss = 0.000000, total_loss = 0.000000 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 150 Avg (IOU: 0.000000), count: 1, class_loss = 0.000000, iou_loss = 0.000000, total_loss = 0.000000 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 161 Avg (IOU: 0.000000), count: 1, class_loss = 0.000000, iou_loss = 0.000000, total_loss = 0.000000 
 total_bbox = 327425, rewritten_bbox = 3.117355 % 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 139 Avg (IOU: 0.000000), count: 31, class_loss = 56.000004, iou_loss = 0.000019, total_loss = 56.000023 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 150 Avg (IOU: 0.000000), count: 18, class_loss = 30.000002, iou_loss = 0.000042, total_loss = 30.000044 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 161 Avg (IOU: 0.000000), count: 4, class_loss = 8.000000, iou_loss = 0.000031, total_loss = 8.000031 
 total_bbox = 327478, rewritten_bbox = 3.118683 %
...
```
Eventually, I get:
```
 detections_count = 0, unique_truth_count = 15222  
class_id = 0, name = person, ap = 0.00%   	 (TP = 0, FP = 0) 

 for conf_thresh = 0.25, precision = -nan, recall = 0.00, F1-score = -nan 
 for conf_thresh = 0.25, TP = 0, FP = 0, FN = 15222, average IoU = 0.00 % 

 IoU threshold = 50 %, used Area-Under-Curve for each unique Recall 
 mean average precision (mAP@0.50) = 0.000000, or 0.00 % 
```

---

The configuration file starts like this:
```
[net]
# Testing
batch=8
subdivisions=8
# Training
batch=8
subdivisions=4
width=736
height=736
channels=3
momentum=0.949
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
burn_in=1000
max_batches = 2000
policy=steps
steps=1600,1800
scales=.1,.1
...
```

Here is a sample of an annotation file (I checked annotations and they should be correct):
```
0 0.6005102040816327 0.8383928571428572 0.017346938775510204 0.06607142857142857
0 0.6362244897959184 0.4107142857142857 0.011224489795918367 0.05357142857142857
0 0.6551020408163265 0.4080357142857143 0.014285714285714285 0.055357142857142855
0 0.6505102040816326 0.3392857142857143 0.015306122448979591 0.05714285714285714
```
Here's `obj.data` file (paths are correct):
```
classes = 1
train = data/train.txt
valid = data/valid.txt
names = data/obj.names
backup = backup/
```

---

If anyone can help, I would appreciate. Thanks !

*PS 1 : I am training on Google Colab. I can't view bounding boxes or mAP curve while training.*

*PS 2 : I got a correct training with the same dataset and an mAP of 10%. Did the owners change something recently (last week) ? Because the last commits didn't seem to be related to training.*

**Comment by nizarmasmoudi** [comment]
It turns out that OpenCV was the problem all along. I used to process (split, mask ignored regions, ...) images using Matplotlib which worked fine. Then, I decided to migrate to OpenCV as was recommended to me by a friend. The same operations gave different results. In fact, OpenCV saves the same image but twice as large in terms of memory and encoding was different than Matplotlib. It may sound weird but OpenCV encoding messed up the training.
Sorry Matplotlib ! I will never betray you again.

**Comment by hichamhendy** [comment]
I also have the same configuration and trained for 2 classes! The accuracy was ugly!!! Did you get a solution?
